CRM/Activity/BAO/Activity.php

```diff
@@ -1763,12 +1765,19 @@ public static function addActivity(
     if ($targetContactID) {
       $activityParams['target_contact_id'][] = $targetContactID;
     }
-    // @todo - use api - remove lots of wrangling above. Remove deprecated fatal & let form layer
-    // deal with any exceptions.
-    if (is_a(self::create($activityParams), 'CRM_Core_Error')) {
-      CRM_Core_Error::fatal("Failed creating Activity for $component of id {$activity->id}");
-      return FALSE;
+
+    // Despite being called "addActivity" this function actually both adds and updates activities
+    // So we have to select the right Api4 action depending on if we have an activity ID.
+    if (!empty($activityParams['id'])) {
+      $activityApi4 = \Civi\Api4\Activity::update();
+    }
+    else {
+      $activityApi4 = \Civi\Api4\Activity::create();
     }
+    $activityApi4
+      ->setValues($activityParams)
+      ->setCheckPermissions(FALSE)
+      ->execute();
```

Comment:
I think this can be simplified to:

```suggestion
    \Civi\Api4\Activity::save()
      ->addRecord($activityParams)
      ->setCheckPermissions(FALSE)
      ->execute();
```